File: services/search-service/src/main/java/com/visor/school/search/service/SearchService.java
package com.visor.school.search.service;

import com.visor.school.search.integration.SearchIndexClient;
import com.visor.school.search.model.SearchIndex;
import com.visor.school.search.model.SearchType;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

/**
 * Search service for indexing and querying documents
 */
@Service
@Slf4j
@RequiredArgsConstructor
public class SearchService {

    private final SearchIndexClient searchIndexClient;

    /**
     * Index a document
     */
    public void index(SearchIndex document) {
        log.info("Indexing document: {} of type: {}", document.getId(), document.getType());
        searchIndexClient.index(document);
    }

    /**
     * Search documents by query and type
     */
    public List<SearchIndex> search(String query, SearchType type) {
        log.info("Searching for: {} in type: {}", query, type);
        return searchIndexClient.search(query, type);
    }

    /**
     * Update an indexed document
     */
    public void updateIndex(SearchIndex document) {
        log.info("Updating indexed document: {}", document.getId());
        searchIndexClient.updateIndex(document);
    }

    /**
     * Delete an indexed document
     */
    public void deleteIndex(UUID id, SearchType type) {
        log.info("Deleting indexed document: {} of type: {}", id, type);
        searchIndexClient.deleteIndex(id, type);
    }
}

File: services/search-service/src/main/kotlin/com/visor/school/search/service/SearchService.kt
package com.visor.school.search.service

import com.visor.school.search.integration.ElasticsearchClient
import com.visor.school.search.model.SearchIndex
import com.visor.school.search.model.SearchType
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.UUID

/**
 * Search service for indexing and querying documents
 */
@Service
class SearchService(
    private val elasticsearchClient: ElasticsearchClient
) {
    private val logger = LoggerFactory.getLogger(SearchService::class.java)

    /**
     * Index a document
     */
    fun index(document: SearchIndex) {
        logger.info("Indexing document: ${document.id} of type: ${document.type}")
        elasticsearchClient.index(document)
    }

    /**
     * Search documents by query and type
     */
    fun search(query: String, type: SearchType): List<SearchIndex> {
        logger.info("Searching for: $query in type: $type")
        return elasticsearchClient.search(query, type)
    }

    /**
     * Update an indexed document
     */
    fun updateIndex(document: SearchIndex) {
        logger.info("Updating indexed document: ${document.id}")
        elasticsearchClient.updateIndex(document)
    }

    /**
     * Delete an indexed document
     */
    fun deleteIndex(id: UUID, type: SearchType) {
        logger.info("Deleting indexed document: $id of type: $type")
        elasticsearchClient.deleteIndex(id, type)
    }
}

